### ISSUE TITLE
LTRIM, RTRIM, BTRIM, TRIM default behavior if chars not given

### ISSUE THREAD

**Issue by terefang** [cockroachdb/cockroach#16043]
**FEATURE REQUEST**

it would be beneficial if the trim functions default charset would behave more like postgresql than using simple 'space' to trim.

* DOCS: http://www.w3resource.com/PostgreSQL/trim-function.php the last example mentions 'whitespace' to trim instead of simple 'space'.

* CONS: SQL-92 specifies only simple 'space' to trim (https://en.wikibooks.org/wiki/SQL_Dialects_Reference/Functions_and_expressions/String_functions)

* PROS: PostgreSQL implements 'whitespace' triming as default (see above).
* PROS: MySQL only mentions 'spaces' in documentation but the implementation uses 'isspace' (https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_trim)

having reviewed the source and go reference, the solution would be to use either:

1. unicode.IsSpace(rune)

2. unicode.IsSpace(rune) || !unicode.IsPrint(rune)

3. unicode.IsSpace(rune) || !unicode.IsPrint(rune) || (rune < 0x20)

i would prefer the 2 or 3 in an unicode enabled world.

**Comment by bdarnell** [comment]
Can you give some examples of strings or characters that are handled differently between postgres and cockroach? It looks like Postgres only removes spaces, not control characters. And what's the difference between your first and second solutions? All runes below `0x20` are non-printable. 

I understand wanting to remove all control or non-printable characters from a string, but it's less clear to me why you'd want to remove only leading or trailing control characters.

**Comment by terefang** [comment]
sorry for my bad formulation of the request and my ignorance of the go language.

having found out that go strings.TrimSpace is equivalent to postgres TRIM implementation makes 50% of the request invalid. 

if all runes below 0x20 are non-printable with go unicode.IsPrint(rune) then i can only argument for solution 1, which would mean to remove leading/trailing garbage (eg hidden control and unicode chars).

it is a common issue in html form field with orm database layers that garbage gets inserted if copy paste is used from other applications with styling features both under windows and linux and also with rdp clipboards.

since we usually cant change the provided applications, being able to use trim() like the following does mitigate the problems greatly:

1. SELECT TRIM(field_with_garbage) FROM ...
2. UPDATE table SET field_with_garbage=TRIM(field_with_garbage)

hope that helps to understand.

**Comment by bdarnell** [comment]
In general we want to be very careful about deviating from PostgreSQL's behavior - if it doesn't remove control characters in TRIM, we probably don't want to either. I recommend researching how you'd remove these undesired characters in PostgreSQL and do the same in CockroachDB (it's still not clear to me exactly which characters are in question here. There are non-printable characters like U+202E RIGHT-TO-LEFT-OVERRIDE which may legitimately appear at the beginning of a string and, if removed, would cause that string to render incorrectly). This may need a new function, and if so we'd be happy to add it. We just don't want to extend widely-implemented functions with different behaviors than they have elsewhere.

**Comment by cuongdo** [comment]
@knz please triage

**Comment by knz** [comment]
After reviewing this I concur we should not alter the behavior of the TRIM functions. Even if CockroachDB supported multiple locales (which it currently doesn't) we would merely restrict the work done by these functions to the whitespace characters defined by the current locale. Otherwise we would be breaking compatibility.

@terefang you can use `regexp_replace` to remove control and non-printable characters. CockroachDB supports more regular expressions than PostgreSQL, they are documented here: https://github.com/google/re2/wiki/Syntax

For example, you can reliably remove control characters at the beginning or the end of a string with:

```sql
regexp_replace(thestring,  '^\p{Cc}*', '', 'g'); -- only beginning
regexp_replace(thestring,  '\p{Cc}*$', '', 'g'); -- only end
regexp_replace(thestring,  '(^\p{Cc}*)|(\p{Cc}*$)', '', 'g'); -- both beginning and end
```

The special regexp sequence "`\p{Cc}`" means "any control character defined by unicode". 

Do you think you could use this in your use case?

**Comment by knz** [comment]
@jseldess we need to highlight in documentation that the regular expressions supported by CockroachDB are much more general than those in PostgreSQL.

**Comment by terefang** [comment]
yes i must admit that if you are aligning with postgresql/sql99 standards
the TRIM functions have to be left as they are.

may i propose an additional optional parameter (true/false, default false)
to enable stripping not only unicode whitespace (Zs,Zl,Zp) but also any
character less than ascii space (0x00-0x20) ?

and how about a global database parameter/variable that can configure this
default (set trim_unicode_mode = true/false)

my use case is different since i have an existing legacy application where
no one wants to change the source.

**Comment by knz** [comment]
If you have a legacy application how come do you need a different behavior. I do not know of any database which strips more than whitespace in these functions.
